zebra-chain/src/mmr.rs

```diff
@@ -0,0 +1,190 @@
+            self.peaks.insert(self.size, entry);
+            self.size += 1;
+        }
+        // TODO: should we really prune every iteration?
+        self.prune()?;
+        // TODO: implement network upgrade logic: drop previous history, start new history
+        Ok(())
+    }
+
+    /// Extend the history tree with the given blocks.
+    pub fn extend<
+        'a,
+        T: IntoIterator<
+            Item = (
+                Arc<Block>,
+                &'a sapling::tree::Root,
+                Option<&'a orchard::tree::Root>,
+            ),
+        >,
+    >(
+        &mut self,
+        iter: T,
+    ) -> Result<(), HistoryTreeError> {
+        for (block, sapling_root, orchard_root) in iter {
+            self.push(block, sapling_root, orchard_root)?;
+        }
+        Ok(())
+    }
+
+    /// Prune tree, removing all non-peak entries.
+    fn prune(&mut self) -> Result<(), io::Error> {
+        // Go through all the peaks of the tree.
+        // This code is from a librustzcash example.
+        // https://github.com/zcash/librustzcash/blob/02052526925fba9389f1428d6df254d4dec967e6/zcash_history/examples/long.rs
+        // See also coins.cpp in zcashd for a explanation of how it works.
+        // https://github.com/zcash/zcash/blob/0247c0c682d59184a717a6536edb0d18834be9a7/src/coins.cpp#L351
+
+        // TODO: should we just copy the explanation here?
+
+        let mut peak_pos_set = HashSet::new();
+        // integer log2 of (self.size+1), -1
+        let mut h = (32 - ((self.size + 1) as u32).leading_zeros() - 1) - 1;
+        let mut peak_pos = (1 << (h + 1)) - 1;
+
+        loop {
+            if peak_pos > self.size {
+                // left child, -2^h
+                peak_pos -= 1 << h;
+                h -= 1;
+            }
+
+            if peak_pos <= self.size {
+                // There is a peak at index `peak_pos - 1`
+                peak_pos_set.insert(peak_pos - 1);
+
+                // right sibling
+                peak_pos = peak_pos + (1 << (h + 1)) - 1;
+            }
+
+            if h == 0 {
+                break;
+            }
+        }
+
+        // Remove all non-peak entries
+        self.peaks.retain(|k, _| peak_pos_set.contains(k));
+        // Rebuild tree
+        self.inner = Tree::new_from_cache(
+            self.network,
+            self.network_upgrade,
+            self.size,
+            &self.peaks,
+            &Default::default(),
+        )?;
+        Ok(())
+    }
+
+    /// Return the hash of the tree root.
+    pub fn hash(&self) -> ChainHistoryMmrRootHash {
+        self.inner.hash()
+    }
+}
+
+impl Clone for HistoryTree {
+    fn clone(&self) -> Self {
+        let tree = Tree::new_from_cache(
+            self.network,
+            self.network_upgrade,
+            self.size,
+            &self.peaks,
+            &HashMap::new(),
+        )
+        .expect("rebuilding an existing tree should always work");
```

Comment:
Testing: Let's also check that `clone` works after a `prune` in the proptests?

Reply:
@teor2345 I'll let you resolve this convo when ready

Reply:
This seems fine for a test.